metrics/metrics.go

```diff
@@ -389,27 +389,74 @@ func GroupHandler() http.Handler {
 	return promhttp.HandlerFor(GroupMetrics, promhttp.HandlerOpts{Registry: GroupMetrics})
 }
 
-// lazyPeerHandler is a structure that defers learning who current
-// group members are until an http request is received.
+// lazyPeerHandler is a structure that defers learning which peers this node is connected
+// to until an http request is received for a specific peer. It handles all peers that
+// this node is connected to regardless of which group they are a part of.
 type lazyPeerHandler struct {
-	peerHandler PeerHandler
+	groupHandlers []GroupHandlers
+	// peerHandlers is a cache of peer address -> handler
+	// TODO Do we need to evict from cache at some point? The only case when it should be
+	//      invalidated is if a peer leaves e.g. during a Reshare
+	peerHandlers sync.Map
+}
+
+// newLazyPeerHandler creates a new lazyPeerHandler from a slice of GroupHandlers
+func newLazyPeerHandler(groupHandlers []GroupHandlers) *lazyPeerHandler {
+	return &lazyPeerHandler{
+		groupHandlers,
+		sync.Map{},
+	}
 }
 
+// handlerForPeer returns the http.Handler associated with a given peer address.
+// Metrics are group-agnostic. Therefore, we just need the Handler for a group that the
+// peer in question has joined, regardless of the group. If a peer belongs to multiple
+// groups, it will return the handler associated with the group that appears first in
+// groupHandlers whose beacon the peer has joined.
+func (l *lazyPeerHandler) handlerForPeer(addr string) (http.Handler, error) {
+	h, found := l.peerHandlers.Load(addr)
+	if found {
+		return h.(http.Handler), nil
+	}
+
+	var handlers map[string]http.Handler
+	var err error
+
+	for _, gh := range l.groupHandlers {
+		handlers, err = gh()
+		if err != nil {
+			// The target addr probably hasn't joined this beacon. Try the next one.
+			continue
+		}
+
+		hdl, found := handlers[addr]
+		if found {
+			l.peerHandlers.Store(addr, hdl)
+			return hdl, nil
+		}
+	}
+
+	return nil, err
```

Comment:
This should probably return an error saying no handler was found, no?

Reply:
good call